sdks/python/apache_beam/runners/interactive/caching/streambasedcache.py

```diff
@@ -0,0 +1,491 @@
+      self._child_subscriptions.append(subscription)
+      if seek_to_start:
+        sub_client.seek(
+            subscription.name,
+            snapshot=self.snapshot.name,
+            retry=api_core.retry.Retry())
+    elif not self._primary_sub_exhausted:
+      subscription = self.subscription
+      self._primary_sub_exhausted = True
+    else:
+      raise ValueError(
+          "Cannot seek_to_start more than once in an environment where "
+          "snapshots are not supported.")
+    return subscription
+
+
+class PatchedPubSubReader(PTransform):
+
+  def __init__(self, cache, reader_class, *reader_args, **reader_kwargs):
+    self._cache = cache
+    self._reader_class = reader_class
+    self._reader_args = reader_args
+    self._reader_kwargs = reader_kwargs
+
+    if "timestamp_attribute" not in reader_kwargs:
+      raise ValueError(
+          "timestamp_attribute must be specified when reading from cache.")
+
+  def expand(self, pbegin):
+    reader_kwargs = self._reader_kwargs.copy()
+
+    reader = self._reader_class(*self._reader_args, **reader_kwargs)
+    if reader_kwargs.get("with_attributes"):
+      return pbegin | reader
+    else:
+      return pbegin | reader | beam.Map(self._cache.coder.decode)
+
+
+class PatchedPubSubWriter(PTransform):
+
+  def __init__(self, cache, writer_class, writer_args, writer_kwargs):
+    self._cache = cache
+    self._writer_class = writer_class
+    self._writer_args = writer_args
+    self._writer_kwargs = writer_kwargs
+
+  def expand(self, pcoll):
+
+    if self._cache.requires_coder and self._cache.coder is None:
+      self._cache.coder = coders.registry.get_coder(pcoll.element_type)
+
+    coder = self._cache.coder
+    writer_kwargs = self._writer_kwargs.copy()
+    attributes_fn = writer_kwargs.pop("attributes_fn", None)
+    # DirectRunner does not support timestamp_attribute
+    timestamp_attribute = writer_kwargs.pop("timestamp_attribute", "ts")
+
+    if writer_kwargs.get("with_attributes"):
+      writer = self._writer_class(*self._writer_args, **writer_kwargs)
+      return pcoll | writer
+    else:
+      # Encode the element as a PubsubMessage ourselves
+      writer_kwargs["with_attributes"] = True
+      encoder = (
+          EncodeToPubSubWithComputedAttributes(
+              coder, attributes_fn, timestamp_attribute) if attributes_fn else
+          EncodeToPubSubWithTimestamp(coder, timestamp_attribute))
+      writer = self._writer_class(*self._writer_args, **writer_kwargs)
+      return pcoll | beam.ParDo(encoder) | writer
+
+
+class DecodeFromPubSub(beam.DoFn):
+
+  def __init__(self, coder, with_attributes, timestamp_attribute):
+    super(DecodeFromPubSub, self).__init__()
+    self.coder = coder
+    self.with_attributes = with_attributes
+    self.timestamp_attribute = timestamp_attribute
+
+  def process(self, message):
+    # pylint: disable=reimported
+    from apache_beam.transforms.window import TimestampedValue
```

Comment:
Could you explain why the process methods have imports in them?

Reply:
Sorry, I have gotten used to doing this when writing DoFns inside scripts / notebooks with `save_main_session=False`.